### PR TITLE
Ensure all paths in `conda.sh` are prefix-replaceable on Windows

### LINKED ISSUE #12626: `conda.sh` has incorrect paths on Windows
### What happened?

Previously, our recipe included a `sed` statement:

https://github.com/conda/conda/blob/ca036ed747d955d23fd0abd83295d96ddf74cc40/recipe/build.sh#L8-L10

This was removed for a simpler recipe implementation in #12603 (which in turn was prompted by issues with packaging conda for Python 3.11, see #12256).

Removing `sed` resulted in the `conda.sh` breaking on Windows + MSYS2. The issue is evident in the first few lines of `conda.sh`:

#### Before Recipe Change (`defaults::conda=23.3.1`)

```bash
export CONDA_EXE='C:/Users/kodegard/.conda/miniconda\Scripts\conda.exe'
export _CE_M=''
export _CE_CONDA=''
export CONDA_PYTHON_EXE='C:/Users/kodegard/.conda/miniconda\python.exe'
```

#### After Recipe Change (`conda-canary/label/dev::conda=23.3.1+23`)

```bash
export CONDA_EXE='/d/a/conda/conda/pkgs/conda_1682642894222/_h_env/Scripts/conda.exe'
export _CE_M=''
export _CE_CONDA=''
export CONDA_PYTHON_EXE='C:/Users/kodegard/.conda/miniconda\python.exe'
```

### PR DESCRIPTION
**Pull request by kenodegard**
### Description

Fix converting Windows path to unix path for cygwin and msys2.

The proposal is to have the PosixActivator conform to what the other activators already do. Within the shell scripts loaded at startup (e.g., the profile) we will rely on `cygpath` (which at that point should be the most correct `cygpath`) to convert our paths. When running `conda activate` we will still perform the path conversion in Python as was done previously.

Resolves #12626

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [ ] ~Add / update outdated documentation?~

### CODE REVIEW

File: conda/activate.py [hunk 1]
```diff
@@ -17,8 +17,9 @@
     isdir,
     join,
 )
+from pathlib import Path
 from textwrap import dedent
-from typing import Iterable
+from typing import Callable, Iterable
```
Comment (travishathaway) [inline conda/activate.py:22]:
```suggestion
from collections.abc import Callable, Iterable
```

File: news/12627-fix-msys2-conda.sh-paths [hunk 2]
```diff
@@ -0,0 +1,19 @@
+### Enhancements
+
+* <news item>
```
Comment (travishathaway) [inline news/12627-fix-msys2-conda.sh-paths:3]: I would also argue that this pull request performs a significant refactor of the way that `Activator` classes are defined in `conda/activate.py`. I would mention this as an enhancement.

**Comment by dbast** [comment]
PR #12509 actually lead to this issue on current main. So either this PR is merged or #12509 reverted as part of next release.

conda.sh is used in "git bash" (part of official windows git installer) and in many cases on GitHub actions windows runners.

**Comment by beeankha** [comment]
pre-commit.ci autofix

**Comment by dbast** [comment]
The title of this PR is a bit misleading... its not about using cygpath but more like **Ensure all paths in conda.sh are prefix replaceable on windows**.

File: news/12627-fix-msys2-conda.sh-paths [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+### Bug fixes
+
+* Stop pre-converting paths to unix style on Windows in `conda.sh`. Instead rely on `cygpath` at runtime. (#12627)
```
Comment (dbast) [inline news/12627-fix-msys2-conda.sh-paths:7]:
```suggestion
* Stop pre-converting paths to unix style on Windows in `conda.sh`, so that they are prefix replaceable upon installation, which got broken by #12509. It also relies on `cygpath` at runtime, which all msys2/cygwin bash versions on windows should have available. (#12627)
```

**Review by dbast — APPROVED** [review]

**Review by beeankha — APPROVED** [review]

File: conda/activate.py [hunk 4]
```diff
@@ -34,7 +35,7 @@
 from .common.path import paths_equal
 
 
-class _Activator:
+class _Activator(metaclass=abc.ABCMeta):
```
Comment (jezdez) [inline conda/activate.py:38]: Woah!

File: conda/activate.py [hunk 5]
```diff
@@ -814,102 +811,83 @@ def ensure_fs_path_encoding(value):
...
     if paths is None:
         return None
-    from subprocess import PIPE, CalledProcessError, Popen
 
-    from conda.auxlib.compat import shlex_split_unicode
+    if not on_win:
+        return path_identity(paths)
+
+    # on windows, uses cygpath to convert windows native paths to posix paths
+    from subprocess import run
```
Comment (jezdez) [inline conda/activate.py:824]: Why is this import here and not at the top of the file?

File: conda/activate.py [hunk 6]
```diff
@@ -814,102 +811,83 @@ def ensure_fs_path_encoding(value):
...
-    return final[0] if single_path else tuple(final)
-
-
-def path_identity(paths):
+def path_identity(paths: str | Iterable[str] | None) -> str | tuple[str] | None:
+    if paths is None:
+        return None
+
     if isinstance(paths, str):
         return os.path.normpath(paths)
```
Comment (jezdez) [inline conda/activate.py:856]:
```suggestion
    if paths is None:
        return None
    elif isinstance(paths, str):
        return os.path.normpath(paths)
```

Comment (jezdez) [reply conda/activate.py:856]: You might combine this with the line below and an else clause, see next suggestion

File: conda/activate.py [hunk 7]
```diff
@@ -814,102 +811,83 @@ def ensure_fs_path_encoding(value):
...
-    elif paths is None:
-        return None
-    else:
-        return tuple(os.path.normpath(_) for _ in paths)
+    return tuple(os.path.normpath(path) for path in paths)
```
Comment (jezdez) [inline conda/activate.py:857]:
```suggestion
    else:
        return tuple(os.path.normpath(path) for path in paths)
```

File: conda/activate.py [hunk 8]
```diff
@@ -814,102 +811,83 @@ def ensure_fs_path_encoding(value):
...
+    if paths is None:
+        return None
 
-        self.unset_var_tmpl = "unset %s"
-        self.export_var_tmpl = "export %s='%s'"
-        self.set_var_tmpl = "%s='%s'"
-        self.run_script_tmpl = '. "%s"'
+    if isinstance(paths, str):
+        return paths.replace("\\", "/")
+    return tuple([path.replace("\\", "/") for path in paths])
```
Comment (jezdez) [inline conda/activate.py:868]:
```suggestion
    elif isinstance(paths, str):
        return paths.replace("\\", "/")
    else:
        return tuple([path.replace("\\", "/") for path in paths])
```

File: conda/activate.py [hunk 9]
```diff
@@ -929,46 +907,40 @@ def _update_prompt(self, set_vars, conda_prompt_modifier):
...
 
-    def _hook_preamble(self):
-        result = ""
+    def _hook_preamble(self) -> str:
+        result = []
         for key, value in context.conda_exe_vars_dict.items():
             if value is None:
                 # Using `unset_var_tmpl` would cause issues for people running
                 # with shell flag -u set (error on unset).
-                # result += join(self.unset_var_tmpl % key) + '\n'
```
Comment (jezdez) [inline conda/activate.py:938]: That line is dropped, any idea why?

File: conda/activate.py [hunk 10]
```diff
@@ -1276,10 +1221,10 @@ def _build_activator_cls(shell):
...
     shell_etc = shell.split("+")
     activator, formatters = shell_etc[0], shell_etc[1:]
-    bases = [activator_map[activator]]
 
+    bases = [activator_map[activator]]
     for f in formatters:
         bases.append(formatter_map[f])
 
-    cls = type("Activator", tuple(bases), {})
+    cls = type("Activator", tuple(reversed(bases)), {})
```
Comment (jezdez) [inline conda/activate.py:1229]: Any reason why the base classes are reversed here?

**Review by jezdez — COMMENTED** [review]
Just some minor questions and nits

File: conda/activate.py (hunk 5 above)
Comment (beeankha) [reply conda/activate.py:824]: I think it's to visually separate where the Windows path-related work is being done, though I'm not 100% sure if that's the reason. If you think it's best to have the import statements at the top of the file, I can move them.

File: conda/activate.py (hunk 9 above)
Comment (dbast) [reply conda/activate.py:938]: What do you mean? That line is a comment line, the next line is converted to append and the `\n` handling moved to the end of the function, which is more generic.

File: conda/activate.py (hunk 10 above)
Comment (jezdez) [reply conda/activate.py:1229]: Gotcha, thanks for elaborating, writing this down for our future selves suffices :)

File: conda/activate.py (hunk 5 above)
Comment (jezdez) [reply conda/activate.py:824]: Generally speaking, I find it a bit of a code smell when imports are allover the place in a module. The only reason to do this is to make sure the module is imported faster. But `subprocess` is probably a module that is loaded during the startup of conda somewhere as well, making this moot.

@jaimergp Have you had any concerns around this working on the very related #12550?

Comment (jezdez) [reply conda/activate.py:824]: For this this isn't a blocker for merging this..

**Review by jezdez — APPROVED** [review]

**Review by beeankha — APPROVED** [review]
